source/upgrade.txt

```diff
@@ -61,17 +59,78 @@ The breaking changes in this section are categorized by the driver version that
+.. example:: Upgrading to Version 4.0
+
+  If you're upgrading the {+driver-short+} from version 3.x to version 4.0,
+  address all breaking changes that are listed for versions 3.x up to and
+  including version 4.0
+
+Types of Breaking Changes
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+New versions of the {+driver-short+} can introduce the following types of breaking changes:
+
+- **API** changes concern the API of the {+driver-short+} libraries. These changes are
+  accompanied by an API major version bump. You might need to edit C++ source code
+  in your project to address these changes.
+
+  .. example::
+
+     - Removes the declaration of function ``v1::func()``. (The definition is still provided for ABI compatibility.)
+     - Removes the inline function ``v_noabi::func()``.
+     - Removes ``v_noabi::func()``.
+     - Changes the redeclaration of ``bsoncxx::foo`` from ``v1::foo`` to ``v2::foo``.
+       To continue using the old interface, qualify ``bsoncxx::foo`` as ``bsoncxx::v1::foo``.
 
-  If you're upgrading the {+driver-short+} from version 3.0 to version 3.10,
-  address all breaking changes listed for versions 3.1 to 3.10, if any.
+- **ABI** changes concern the ABI of the {+driver-short+} libraries. These changes are
+  accompanied by an ABI version bump, and may also be accompanied by an API major version
+  bump. You might need to rebuild your project or edit your C++ source code to address
+  these changes.
 
+  ABI stability is **not** supported for symbols corresponding to entities declared in
+  the ``v_noabi`` namespace. Breaking changes for such symbols are not necessarily
+  accompanied by an ABI version bump.
+
+  .. example::
+
+     - Removes the exported symbol ``v_noabi::func()``.
+     - Removes the exported symbol ``v1::func()``.
+     - Removes the exported symbols for ``v1::type``.
+
+- **Build system** changes concern the build system or packaging of the {+driver-short+}
+  libraries. These changes are not necessarily accompanied by either an API major version
+  bump or an ABI version bump.
+  To address these changes, you might need to modify your build system configurations or
+  update your project's package or dependency management settings. You might also need to
+  modify or rebuild your C++ source code.
+
+  .. example::
+
+     - Removes the ``ENABLE_FOO`` configuration variable.
+     - Changes the default value of the ``ENABLE_BAR`` configuration variable from
+       ``OFF`` to ``ON``.
+     - Drops support for platform A, compiler B, or architecture C.
+
+.. _cpp-breaking-changes:
+
+Breaking Changes
+----------------
+
+The following sections describe the breaking changes introduced by 
+each driver version release.
+
+.. _upcoming-breaking-changes:
 .. _version-4.0-breaking-changes:
 
-Version 4.0 Breaking Changes
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Version 4.0
+~~~~~~~~~~~
+
+.. _version-4.0-build-system-breaking:
 
-This driver version introduces the following breaking changes:
+Build System Breaking Changes
```

Comment:
This new section layout looks fantastic, but I foresee a problem with link stability. It seems like the current subsection headers with the same name are being given numeric suffixes by order of instance:

- `https://.../upgrade/#build-system-breaking-changes` (v4.0)
- `https://.../upgrade/#build-system-breaking-changes-1` (v3.11)
- `https://.../upgrade/#build-system-breaking-changes-2` (v3.10)
- ...

This means when a new entry is added to the page, all prior links to individual subsections of a given version will be incorrectly offset by one and refer to the section for a _newer_ version than intended.

The addition of version-specific subsection "anchors"(?) may be needed to preserve link stability. If that does not suffice, the subsection names may need to be made unique (although I would prefer keeping their current names as-is if possible).

Reply:
Oh I see, thanks for pointing this out! Unfortunately adding specific anchors doesn't change the link names, so I added the version to each subsection name